### ISSUE TITLE
Google Play Billing  5.0.0，queryProductDetails can not get product details

### ISSUE THREAD

**Issue by zcqzcq** [android/play-billing-samples#563]
Response Code: FEATURE_NOT_SUPPORTED, Debug Message: Client does not support ProductDetails.

**Comment by chengxuyuan12345** [comment]
Just update the play shop

**Comment by huize** [comment]
> Just update the play shop

we can not request all users to update google play store

**Comment by AtlantisFaded** [comment]
i tried to update play store, but it shows it's the latest. And ??? I have nothing to do.

**Comment by chengxuyuan12345** [comment]
> 

api does not support older versions no way

**Comment by Ayagikei** [comment]
The ProductDetails attribute is relative to the version of the Google Play Store installed by the user.

ProductDetails is unfortunately not backward compatible.

**You can use the following methods to determine whether the user supports ProductDetails, and prompt the user:**
``` kotlin
val isProductDetailsSupported =
            billingClient.isFeatureSupported(BillingClient.FeatureType.PRODUCT_DETAILS).responseCode == BillingClient.BillingResponseCode.OK
````

If you don't need the new features of ProductDetails, you can actually use SkuDetails or make a layer of compatibility, for example:
```kotlin
sealed class ProductDetailsCompact {
    class ProductDetailType(val productDetails: ProductDetails) :
        ProductDetailsCompact(isFromCached)

    class SkuDetailType(val skuDetails: SkuDetails) :
        ProductDetailsCompact(isFromCached)
}
````

But this also means that both query items and launchBillingFlow need to do two kinds of processing.

**Comment by stephano1215** [comment]
Is there any document that describe the standard of Play Store version for Google Play Billing Library 5.0.0? 
Some of users can not access to product details even though their Play Store version is up to date.

**Comment by NameX44** [comment]
Hi, same problem here, user's Play Store is up to date but they can't get product details anyway.
I've got the same problem on an old emulator API 30 with Play Store up to date too.

**Comment by jacob-lerenius** [comment]
This is a terrible update from Google, Google Play says it is the latest version on some devices but the product details are not supporten. When using the deprecated sku details API, there is a response saying the Google Play store need to be updated. This will cause severe issues if released. The new API should of course fallback to the old API for older versions automatically.

**Comment by jacob-lerenius** [comment]
> @jacob-lerenius does this message appear only on some devices? I reverted to sku details and it is working when i tested
@renanrbs Yes, on some devices. Google Play says it is the latest version, yet, it does not work and it complains on the version. This is not going to work well for many users when releasing. The device is about 5 years old. I tried to revert back to Billing library V4 but it still complains, so SDK version probably also matters as I updated to latest SDK version as well. Needs some more investigation.

**Comment by vast00** [comment]
> > @jacob-lerenius does this message appear only on some devices? I reverted to sku details and it is working when i tested
> > @renanrbs Yes, on some devices. Google Play says it is the latest version, yet, it does not work and it complains on the version. This is not going to work well for many users when releasing. The device is about 5 years old. I tried to revert back to Billing library V4 but it still complains, so SDK version probably also matters as I updated to latest SDK version as well. Needs some more investigation.

Hi, same error here.
Is there any any progress?

**Comment by saryongkang** [comment]
> 

Could you elaborate on this? The feature `ProductDetails` is supported by Play Store ver. 26.4 (July, 2021) and up. Your users still experience same issue even if they updated Play Store to the latest one, please file a issue with more details so that we can reproduce it: https://issuetracker.google.com/issues/new?component=190923

Also, we recommend that you call `isFeatureSupported(PRODUCT_DETAILS)` to determine if Play Store supports the feature on the current device, and then 1) either fall back to `querySkuDetailsAsync()` method or 2) prompt the user to update their Play Store to avoid future purchase errors.

**Comment by jacob-lerenius** [comment]
> > > @jacob-lerenius does this message appear only on some devices? I reverted to sku details and it is working when i tested
> > > @renanrbs Yes, on some devices. Google Play says it is the latest version, yet, it does not work and it complains on the version. This is not going to work well for many users when releasing. The device is about 5 years old. I tried to revert back to Billing library V4 but it still complains, so SDK version probably also matters as I updated to latest SDK version as well. Needs some more investigation.
> 
> Hi, same error here. Is there any any progress?
I think the main issue in this case is that Google Play Store gets stuck between not finding an update but require a later version when launching a purchase flow.

**Comment by jacob-lerenius** [comment]
> > This is a terrible update from Google, Google Play says it is the latest version on some devices but the product details are not supported. When using the deprecated sku details API, there is a response saying the Google Play store need to be updated when launching the purchas flow. This will cause severe issues if released. The new API should of course fallback to the old API for older versions automatically.
> 
> @jacob-lerenius Could you elaborate on this? The feature `ProductDetails` is supported by Play Store ver. 26.4 (July, 2021) and up. Your users still experience same issue even if they updated Play Store to the latest one, please file a issue with more details so that we can reproduce it: https://issuetracker.google.com/issues/new?component=190923
> 
> We recommend that you call `isFeatureSupported(PRODUCT_DETAILS)` to determine if Play Store supports the feature on the current device, and then either 1) fall back to `querySkuDetailsAsync()` method or 2) prompt the user to update their Play Store to avoid future purchase errors.

I understand the workaround even though it is a really ugly update by the Billing team. The probem I see is that the Google Play Store complains that the version needs to be updated when launching a puchase flow, but at the same time says it is the latest version when checking for updates in the Google Play Store app. It is stuck there. No idea how common this is.

**Comment by renanrbs** [comment]
Hey @saryongkang, there is already a issue registered [here](https://issuetracker.google.com/issues/256023689). I believe the problem happens only for some users, so i couldn't find a way to reproduce it. But, i've followed the suggestion to use the `isFeatureSupported` method and it seems to have solved the issue.
I believe the ideal solution would be to not need that check since it is not a optional feature, you can't [use the billing library](https://developer.android.com/google/play/billing/integrate#launch) without retrieving the product details and it looks really ugly to include deprecated apis in our code, but, if that is the only way to do it, it should be in the [migration guide](https://developer.android.com/google/play/billing/migrate-gpblv5) to avoid further problems

**Comment by jacob-lerenius** [comment]
> Hey @saryongkang, there is already a issue registered [here](https://issuetracker.google.com/issues/256023689). I believe the problem happens only for some users, so i couldn't find a way to reproduce it. But, i've followed the suggestion to use the `isFeatureSupported` method and it seems to have solved the issue. I believe the ideal solution would be to not need that check since it is not a optional feature, you can't [use the billing library](https://developer.android.com/google/play/billing/integrate#launch) without retrieving the product details and it looks really ugly to include deprecated apis in our code, but, if that is the only way to do it, it should be in the [migration guide](https://developer.android.com/google/play/billing/migrate-gpblv5) to avoid further problems

If this is the suggested solution, the API should not be deprecated and removed in a future release.

**Comment by marcardar** [comment]
Just had a report from a device (Galaxy A12, Android 11) running the latest versions of both Google Play Services (version code: `224814037`) and Google Play Store (version name: `33.5.17-21 [0] [PR] 493922335`) where the billing response code for `billingClient.isFeatureSupported(BillingClient.FeatureType.PRODUCT_DETAILS)` was `FEATURE_NOT_SUPPORTED`. How is this possible? Perhaps unsupported country? Next, I'll add network country code to the reporting.

**Comment by mahdi-rovio** [comment]
I have the same issue with Play Store app version 34.0.12.-21 [0] [PR] 500233025 in a Galaxy S211 Android 13.
There is no update for the Play Store app.
Clicking the "Update Play Store" button in Play Store app's Settings (About section) reports: "Google Play Store is up to Date"

**Comment by Shyrou** [comment]
Hi there, I hope this helps someone.

I was having the same issue until I tried the following:

- Go to Settings > Applications > Google Play Store
- In Storage section, delete de Cache and also the Data
- Force close Google Play Store (not sure this step is necessary)
- Open your app and try to request product details again

Please let me know if this works for you, and also if you needed to force close Google Play Store or if deleting the data was enough. Thanks!

**Comment by dogtim** [comment]
It still happens after clearing the data and cache in Google Play Store App.
Here is the list of devices that encounters the problem.

Android 9, Samsung Galaxy S8+
Android 12, Samsung Galaxy A32
Android 13, Samsung Galaxy S21 Ultra
Android 13, Samsung Galaxy S20 FE
Android 13, Samsung Galaxy S20 FE
Android 13, Samsung Galaxy S20

Could we get the source code for the Billing Client Library? We want to check the call stack and find the root cause of why some devices have this problem

**Comment by mahdi-hamdarsi-r** [comment]
@Shyrou Tried that. Didn't workout. It got fixed two days later though, with no apparent solution.

**Comment by claudioredi** [comment]
I also get a bunch of FEATURE_NOT_SUPPORTED when executing queryProductDetails. Many on them are on Android 13 or Android 12... how is that possible? I assume Google Play Store version won't be very old. It's happening in the wild so I don't have that information but I was expecting this problem to affect devices running old Android versions.

**Comment by Gropoid** [comment]
This error was driving me crazy, on some devices but not others. 
It turns out that it's returned if your account in Google Play Store does not have a means of payment setup. 
So, for me the fix was to use an account with a Credit Card linked to it.
I wish the error code were more explicit.

**Comment by marcardar** [comment]
@Gropoid I also experienced this from a user who had all the latest Play Store app and Play Services versions as well as cleared caches. Eventually we discovered they had some other Google account (presumably without payment setup) that was being used by the Play Store. A more explicit message would make sense and be very useful.
